modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/http/conn/ProxyConfig.java

```diff
@@ -85,6 +116,196 @@ public HttpHost selectProxy(final HttpHost target) {
+     */
+    public UsernamePasswordCredentials getCredentialsForTargetHost(String targetHost) {
+
+        UsernamePasswordCredentials credentials = null;
+        ProxyProfileConfig proxyProfileForTargetHost = getProxyProfileForTargetHost(targetHost);
+
+        if (proxyProfileForTargetHost != null) {
+            credentials = proxyProfileForTargetHost.getCredentials();
+        }
+
+        return credentials;
+    }
+
+    /**
+     * returns appropriate log message based on the proxy configuration
+     * whether loading proxy profile or single proxy server or no proxy configured
+     *
+     * @return log message
+     */
+    public String logProxyConfig() {
+        if (isProxyProfileConfigured()) {
+            return "HTTP Sender using proxy profile";
+        }
+
+        if (this.proxy != null) {
+            return "HTTP Sender using Proxy " + getProxy() + " and  bypassing " + getProxyBypass();
+        } else {
+            return "No proxy configuration found";
+        }
+    }
+
+    /**
+     * checks whether proxy configured (either proxy profile or single server)
+     *
+     * @return true when either proxy profile is configure or default proxy server is configure, false otherwise
+     */
+    private boolean isProxyConfigured() {
+        return proxy != null || isProxyProfileConfigured();
+    }
+
+    /**
+     * checks the proxy configuration and profile configuration whether proxy is configured with credential
+     *
+     * @return true when at least one proxy has configured with credential, false otherwise
+     */
+    private boolean isProxyHasCredential() {
+        if (!isProxyConfigured()) {
+            return false;
+        }
+
+        if (!isProxyProfileConfigured()) {
+            return getCreds() != null;
+        }
+
+        for (Map.Entry<String, ProxyProfileConfig> proxyProfile : this.proxyProfileMap.entrySet()) {
+            if (proxyProfile.getValue().getCredentials() != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * returns DefaultProxyAuthenticator if single proxy server is configured
+     * ProfileProxyAuthenticator if proxy profile is configured
+     *
+     * @return ProxyAuthenticator, if proxy is not configured null
+     * @throws AxisFault
+     */
+    public ProxyAuthenticator createProxyAuthenticator() throws AxisFault {
+        if (!isProxyHasCredential()) return null;
+
+        ProxyAuthenticator proxyAuthenticator;
+
+        try {
+            if (isProxyProfileConfigured()) {
+                proxyAuthenticator = new ProfileProxyAuthenticator(this);
+            } else {
+                proxyAuthenticator = new DefaultProxyAuthenticator(getCreds());
+            }
+
+        } catch (MalformedChallengeException e) {
+            throw new AxisFault("Error while creating proxy authenticator", e);
+        }
+
+        return proxyAuthenticator;
+    }
+
     @Override
     public String toString() {
```

Comment:
Although this is not a part of your pull request, this toString() method looks overly complicated.